File: backend/platform-common/src/main/java/com/selfcare/platform/common/context/RequestContext.java
package com.selfcare.platform.common.context;

import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Canonical request context propagated across every call in the platform.
 *
 * Immutable, industry-neutral, and populated from incoming headers (see {@link #fromHeaders(Map)}).
 * The same context travels through the api-gateway (reactive), the servlet services, and the
 * async/kafka processing chains via {@link RequestContextHolder}, {@link RequestContextWebFilter}
 * and {@link RequestContextTaskDecorator}.
 *
 * <p>Header contract (SDK mirrors these exact names):</p>
 * <ul>
 *   <li>X-Tenant-Id — registered tenant (validated)</li>
 *   <li>X-User-Id — authenticated user</li>
 *   <li>X-Session-Id — session token reference</li>
 *   <li>X-Correlation-Id — request correlation</li>
 *   <li>X-Trace-Id — distributed trace id</li>
 *   <li>X-Channel — MOBILE / WEB / API / KIOSK / IVR / PARTNER</li>
 *   <li>X-Device-Id — originating device</li>
 *   <li>Accept-Language — requested locale (e.g. en, en-SG)</li>
 *   <li>X-Forwarded-For — caller ip</li>
 *   <li>X-Environment — dev / qa / staging / prod</li>
 * </ul>
 */
public record RequestContext(
        String tenantId,
        String userId,
        String userType,
        String sessionId,
        String token,
        String correlationId,
        String traceId,
        String channel,
        String deviceId,
        String ipAddress,
        Locale locale,
        String environment,
        String authMethod) {

    public static final String TENANT_HEADER = "X-Tenant-Id";
    public static final String USER_HEADER = "X-User-Id";
    public static final String SESSION_HEADER = "X-Session-Id";
    public static final String CORRELATION_HEADER = "X-Correlation-Id";
    public static final String TRACE_HEADER = "X-Trace-Id";
    public static final String CHANNEL_HEADER = "X-Channel";
    public static final String DEVICE_HEADER = "X-Device-Id";
    public static final String LOCALE_HEADER = "Accept-Language";
    public static final String FORWARDED_FOR_HEADER = "X-Forwarded-For";
    public static final String ENVIRONMENT_HEADER = "X-Environment";

    public static RequestContext empty() {
        return new RequestContext(
                "UNKNOWN", null, null, null, null, null, null, null, null, null, null, null, null);
    }

    public static RequestContext with(String tenantId, String correlationId) {
        return new RequestContext(
                tenantId, null, null, null, null, correlationId, null, null, null, null, null, null, null);
    }

    public static RequestContext fromHeaders(Map<String, String> headers) {
        return new RequestContext(
                value(headers, TENANT_HEADER),
                value(headers, USER_HEADER),
                value(headers, "X-User-Type"),
                value(headers, SESSION_HEADER),
                value(headers, "Authorization"),
                value(headers, CORRELATION_HEADER),
                value(headers, TRACE_HEADER),
                value(headers, CHANNEL_HEADER),
                value(headers, DEVICE_HEADER),
                value(headers, FORWARDED_FOR_HEADER),
                parseLocale(value(headers, LOCALE_HEADER)),
                value(headers, ENVIRONMENT_HEADER),
                null);
    }

    /** Materialize this context as the canonical outgoing header map (SDK and downstream calls). */
    public Map<String, String> toHeaders() {
        Map<String, String> headers = new LinkedHashMap<>();
        if (tenantId != null) headers.put(TENANT_HEADER, tenantId);
        if (userId != null) headers.put(USER_HEADER, userId);
        if (userType != null) headers.put("X-User-Type", userType);
        if (sessionId != null) headers.put(SESSION_HEADER, sessionId);
        if (token != null) headers.put("Authorization", token);
        if (correlationId != null) headers.put(CORRELATION_HEADER, correlationId);
        if (traceId != null) headers.put(TRACE_HEADER, traceId);
        if (channel != null) headers.put(CHANNEL_HEADER, channel);
        if (deviceId != null) headers.put(DEVICE_HEADER, deviceId);
        if (ipAddress != null) headers.put(FORWARDED_FOR_HEADER, ipAddress);
        if (locale != null) headers.put(LOCALE_HEADER, locale.toLanguageTag());
        if (environment != null) headers.put(ENVIRONMENT_HEADER, environment);
        if (authMethod != null) headers.put("X-Auth-Method", authMethod);
        return headers;
    }

    private static String value(Map<String, String> headers, String name) {
        String v = headers.get(name);
        return v != null && !v.isBlank() ? v.trim() : null;
    }

    private static Locale parseLocale(String tag) {
        if (tag == null || tag.isBlank()) return null;
        try {
            return Locale.forLanguageTag(tag.split(",")[0].trim());
        } catch (RuntimeException ex) {
            return null;
        }
    }
}

File: admin/selfcare-admin/src/services/requestContext.ts
/**
 * RequestContext SDK — web (selfcare Studio) side of the canonical context contract.
 *
 * Mirrors `com.selfcare.platform.common.context.RequestContext` in platform-common and the
 * mobile SDK (`mobile/selfcare-app/src/services/RequestContext.ts`). Same header contract,
 * same lifecycle: reset on login, update on tenant switch, apply to every API request.
 */

export type RequestChannel = 'MOBILE' | 'WEB' | 'API' | 'KIOSK' | 'IVR' | 'PARTNER';

export interface RequestContextSpec {
  tenantId?: string;
  userId?: string;
  userType?: string;
  sessionId?: string;
  token?: string;
  correlationId?: string;
  traceId?: string;
  channel?: RequestChannel;
  deviceId?: string;
  ipAddress?: string;
  locale?: string;
  environment?: 'dev' | 'qa' | 'staging' | 'prod';
  authMethod?: string;
}

export const REQUEST_HEADERS = {
  TENANT: 'X-Tenant-Id',
  USER: 'X-User-Id',
  USER_TYPE: 'X-User-Type',
  SESSION: 'X-Session-Id',
  CORRELATION: 'X-Correlation-Id',
  TRACE: 'X-Trace-Id',
  CHANNEL: 'X-Channel',
  DEVICE: 'X-Device-Id',
  LOCALE: 'Accept-Language',
  FORWARDED_FOR: 'X-Forwarded-For',
  ENVIRONMENT: 'X-Environment',
  AUTH_METHOD: 'X-Auth-Method',
} as const;

export type RequestContextHeaders = Record<string, string>;

const CORRELATION_PREFIX = 'selfcare-';

export function newCorrelationId(): string {
  const rand = Math.random().toString(16).slice(2, 10);
  return `${CORRELATION_PREFIX}${rand}`;
}

class RequestContext {
  private static instance: RequestContext | null = null;
  private spec: RequestContextSpec;

  private constructor(initial?: RequestContextSpec) {
    this.spec = { correlationId: newCorrelationId(), ...(initial ?? {}) };
  }

  static current(initial?: RequestContextSpec): RequestContext {
    if (!RequestContext.instance) {
      RequestContext.instance = new RequestContext(initial);
    }
    return RequestContext.instance;
  }

  static reset(initial?: RequestContextSpec): RequestContext {
    RequestContext.instance = new RequestContext(initial);
    return RequestContext.instance;
  }

  update(patch: RequestContextSpec): this {
    this.spec = { ...this.spec, ...patch };
    return this;
  }

  get tenantId(): string | undefined {
    return this.spec.tenantId;
  }

  headers(): RequestContextHeaders {
    const h: RequestContextHeaders = {};
    if (this.spec.tenantId) h[REQUEST_HEADERS.TENANT] = this.spec.tenantId;
    if (this.spec.userId) h[REQUEST_HEADERS.USER] = this.spec.userId;
    if (this.spec.userType) h[REQUEST_HEADERS.USER_TYPE] = this.spec.userType;
    if (this.spec.sessionId) h[REQUEST_HEADERS.SESSION] = this.spec.sessionId;
    if (this.spec.traceId) h[REQUEST_HEADERS.TRACE] = this.spec.traceId;
    if (this.spec.channel) h[REQUEST_HEADERS.CHANNEL] = this.spec.channel;
    if (this.spec.deviceId) h[REQUEST_HEADERS.DEVICE] = this.spec.deviceId;
    if (this.spec.locale) h[REQUEST_HEADERS.LOCALE] = this.spec.locale;
    if (this.spec.ipAddress) h[REQUEST_HEADERS.FORWARDED_FOR] = this.spec.ipAddress;
    if (this.spec.environment) h[REQUEST_HEADERS.ENVIRONMENT] = this.spec.environment;
    if (this.spec.authMethod) h[REQUEST_HEADERS.AUTH_METHOD] = this.spec.authMethod;
    h[REQUEST_HEADERS.CORRELATION] = this.spec.correlationId ?? newCorrelationId();
    if (this.spec.token) {
      h['Authorization'] = this.spec.token.startsWith('Bearer ')
        ? this.spec.token
        : `Bearer ${this.spec.token}`;
    }
    return h;
  }

  applyTo(headers: Record<string, string>): Record<string, string> {
    return { ...this.headers(), ...headers };
  }

  snapshot(): RequestContextSpec {
    return { ...this.spec };
  }
}

export function requestContext(): RequestContext {
  return RequestContext.current();
}

export function requestContextHeaders(): RequestContextHeaders {
  return RequestContext.current().headers();
}

export function withRequestContext(headers: Record<string, string>): Record<string, string> {
  return RequestContext.current().applyTo(headers);
}